### PR TITLE
[Baremetal] Add API + Stub implementation for DTLS Connection ID

### PR DESCRIPTION
**Pull request by hanno-arm**
__Context/Problem:__ When running a long-lived DTLS connection, there is no stable IP+Port metadata through which the connection can consistently be identified. Therefore, for every participant in such a long-lived connection which has more than one DTLS connection open a means must be provided to associate an incoming DTLS record to a DTLS session.

__Connection ID extension:__ The [DTLS Connection ID](https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-04#section-4) extension allows to reliably associate DTLS records to DTLS connections across changes in the underlying transport by adding explicit connection identifiers (CIDs) to the headers of encrypted DTLS records. The desired CIDs are configured by the application layer and are exchanged in new `ClientHello` / `ServerHello` extensions during the handshake, where each side indicates the CID it wants the peer to use when writing encrypted messages. The CIDs are put to use once records get encrypted: the stack discards any incoming records that don't include the configured CID in their header, and adds the peer's requested CID to the headers of outgoing messages.

__Summary:__ This is the first in a series of PRs adding support for the DTLS Connection ID extension. It (a) adds and documents a new compile-time options and API to configure the use of the Connection ID extension, (b) adds stub implementations for the new API, (c) exemplifies their use in the `ssl_client2` and `ssl_server2` example applications, (d) adds stub tests to `ssl-opt.sh`.

__API/ABI changes:__ This PR suggests the addition of the compile-time option `MBEDTLS_SSL_CID`, disabled by default, which controls the availability of additional SSL API allowing to use the CID extension with Mbed TLS. This additional SSL API consists of the pre-handshake function `mbedtls_ssl_set_cid()`, which allows to enable/disable the use of the CID extension at compile time, and  the post-handshake function `mbedtls_ssl_get_peer_cid()`, which allows to check if the use of the CID has been negotiated. The PR also suggests the addition of compile-time constants `MBEDTLS_SSL_CID_IN_MAX_LEN, MBEDTLS_SSL_CID_OUT_MAX_LEN` controlling the maximum lengths of CIDs at compile-time.
The design of the implementation isn't complete, yet. It is expected that the implementation will break the ABI. It is expected that it does not modify existing API.

__Worklist:__
- [x] Add and document Connection ID specific compile-time configurations
- [x] Add and document Connection ID specific API extensions
- [x] Add dummy implementations of Connection ID API
- [x] Add command line options to control the use of Connection ID in `ssl_client2` and `ssl_server2`
- [x] Extend `ssl-opt.sh` integration tests with basic, single-threaded tests for Connection ID

Later PRs will take care of:
- Implement parsing/writing of Connection ID extensions
- Implement CID-specific record encryption/decryption for AEAD modes
- Add unit tests for CID-specific record encryption/decryption for AEAD modes
- Implement CID-specific record encryption/decryption for non-AEAD modes
- Add unit tests for CID-specific record encryption/decryption for non-AEAD modes
- Adapt SSL record layer to be able to deal with change in record type

### CODE REVIEW

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -1396,6 +1410,123 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+ *
+ *                    This API allows to enable/disable the use of the CID
+ *                    extension in the next handshake and to set the value of
+ *                    the CID to be used for incoming messages.
+ *
+ * \param ssl         The SSL context to configure. This must be initialized.
+ * \param enable      This value determines whether the CID extension should
+ *                    be used or not. Possible values are:
+ *                    - MBEDTLS_SSL_CID_ENABLED to enable the use of the CID.
+ *                    - MBEDTLS_SSL_CID_DISABLED to disable the use of the CID.
```
Comment (mpg) [inline include/mbedtls/ssl.h]: I assume disabled is the default value (ie the one used if the function is never called), but it would probably be worth stating explicitly.

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -1396,6 +1410,123 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+ * \param own_cid     The address of the readable buffer holding the CID we want
+ *                    the peer to use when sending encrypted messages to us.
+ *                    This may be \c NULL if \p own_cid_len is \c 0.
+ *                    This parameter is unused if \p enabled is set to
+ *                    MBEDTLS_SSL_CID_DISABLED.
+ * \param own_cid_len The length of \p own_cid.
+ *                    This parameter is unused if \p enabled is set to
+ *                    MBEDTLS_SSL_CID_DISABLED.
+ *
+ * \note              This CID configuration applies to the next handshake
```
Comment (mpg) [inline include/mbedtls/ssl.h]: Does it apply only to the next one, or all subsequent ones until further calls to this function are made? For example if I call this function then `handshake()` then `renegotiate()`, is the CID set by the initial call to this function used for the renegotiating handshake as well?

File: include/mbedtls/check_config.h [hunk 3]
```diff
@@ -591,6 +591,11 @@
 #error "MBEDTLS_SSL_DTLS_ANTI_REPLAY  defined, but not all prerequisites"
 #endif
 
+#if defined(MBEDTLS_SSL_CID) &&                              \
+    ( !defined(MBEDTLS_SSL_TLS_C) || !defined(MBEDTLS_SSL_PROTO_DTLS) )
+#error "MBEDTLS_SSL_CID  defined, but not all prerequisites"
+#endif
+
```
Comment (mpg) [inline include/mbedtls/check_config.h]: Note on commit structure: in the future, we should get into the habit that every time we add something new to `config.h` in the same commit we:

- update generated files (`version_feature`, `query_config`)
- update `check_config.h` if necessary

Otherwise the commit isn't quite atomic but rather subatomic.

(That's just a note for the future though, note a request to rewrite history here.)

File: programs/ssl/ssl_server2.c [hunk 4]
```diff
@@ -2786,6 +2850,37 @@ int main( int argc, char *argv[] )
...
+            mbedtls_printf( " failed\n  ! mbedtls_ssl_get_peer_cid returned -0x%x\n\n",
+                            -ret );
+            goto exit;
+        }
+
+        if( cid_negotiated == MBEDTLS_SSL_CID_DISABLED )
+        {
+            if( opt.cid_enabled == MBEDTLS_SSL_CID_ENABLED )
+            {
+                mbedtls_printf( "Use of Connection ID was rejected by the client.\n" );
```
Comment (mpg) [inline programs/ssl/ssl_server2.c]: Nitpicking: "not offered" would be more precise than "rejected". As a server, we don't get to be rejected because we can't even offer things.

File: programs/ssl/ssl_server2.c [hunk 5]
```diff
@@ -2786,6 +2850,37 @@ int main( int argc, char *argv[] )
...
+            }
+        }
+        else
+        {
+            mbedtls_printf( "Use of Connection ID has been negotiated.\n" );
```
Comment (mpg) [inline programs/ssl/ssl_server2.c]: I think we should also print the value of the negotiated ID here for testing purposes (if only to make sure `get_peer_cid()` works properly).

File: programs/ssl/ssl_client2.c [hunk 6]
```diff
@@ -1689,6 +1755,37 @@ int main( int argc, char *argv[] )
...
+        if( cid_negotiated == MBEDTLS_SSL_CID_DISABLED )
+        {
+            if( opt.cid_enabled == MBEDTLS_SSL_CID_ENABLED )
+            {
+                mbedtls_printf( "Use of Connection ID was rejected by the server.\n" );
+            }
+        }
+        else
+        {
+            mbedtls_printf( "Use of Connection ID has been negotiated.\n" );
```
Comment (mpg) [inline programs/ssl/ssl_client2.c]: I think we should also print the value of the negotiated ID here for testing purposes (if only to make sure `get_peer_cid()` works properly).

File: tests/ssl-opt.sh [hunk 7]
```diff
@@ -1112,6 +1112,103 @@ run_test    "Truncated HMAC, DTLS: client enabled, server enabled" \
...
+# Tests for DTLS Connection ID extension
+
+# TODO
+# So far, the CID API isn't implemented, so we can't
+# grep for output witnessing its use. This needs to be
+# changed once the CID extension is implemented.
+
+requires_config_enabled MBEDTLS_SSL_CID
+run_test    "Connection ID: Client enabled, server disabled" \
+            "$P_SRV dtls=1cid=0" \
```
Comment (mpg) [inline tests/ssl-opt.sh]: Missing space.

File: tests/ssl-opt.sh [hunk 8]
```diff
@@ -1112,6 +1112,103 @@ run_test    "Truncated HMAC, DTLS: client enabled, server enabled" \
...
+            0
+
+requires_config_enabled MBEDTLS_SSL_CID
+run_test    "Connection ID: Client+Server enabled, Client+Server CID empty" \
+            "$P_SRV dtls=1 cid=1" \
+            "$P_CLI dtls=1 cid=1" \
+            0
+
+requires_config_enabled MBEDTLS_SSL_CID
+run_test    "Connection ID: Client+Server enabled, Client+Server CID nonempty, AES-128-CCM-8" \
```
Comment (mpg) [inline tests/ssl-opt.sh]: I'm not sure I get why forcing a cipher is useful here. (Same with the CBC thing below.)

File: tests/ssl-opt.sh [hunk 9]
```diff
@@ -1112,6 +1112,103 @@ run_test    "Truncated HMAC, DTLS: client enabled, server enabled" \
...
+            "$P_CLI dtls=1 cid=1 cid_val=deadbeef force_ciphersuite=TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256" \
+            0
+
+requires_config_enabled MBEDTLS_SSL_CID
+run_test    "Connection ID: Client+Server enabled, Client+Server CID empty, AES-128-CBC" \
+            "$P_SRV dtls=1 cid=1" \
+            "$P_CLI dtls=1 cid=1 force_ciphersuite=TLS-ECDHE-ECDSA-WITH-AES-128-CBC-SHA256" \
+            0
+
+requires_config_enabled MBEDTLS_SSL_CID MBEDTLS_SSL_RENEGOTIATION
```
Comment (mpg) [inline tests/ssl-opt.sh]: This function only takes one argument, you probably want to call it twice instead. (Though perhaps having it take multiple argument would be nicer, but changing that is probably out of scope here.)

File: library/ssl_tls.c [hunk 10]
```diff
@@ -109,6 +109,33 @@ static void ssl_update_in_pointers( mbedtls_ssl_context *ssl,
 
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
 
+/* Top-level Connection ID API */
+
+int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
```
Comment (mpg) [inline library/ssl_tls.c]: Please add a big fat warning to the documentation of the function that so far it does nothing (and always returns success). We want this branch to be mergeable, so that people who want to start coding against that API (as the stub functions aim to allow) can do it from the main branch (they might need other features from the main branch and shouldn't have to to there own merges), but I don't feel comfortable merging a stub implementation without it being documented.

File: library/ssl_tls.c [hunk 11]
```diff
@@ -109,6 +109,33 @@ static void ssl_update_in_pointers( mbedtls_ssl_context *ssl,
...
+                         size_t own_cid_len )
+{
+    ((void) ssl);
+    ((void) enable);
+    ((void) own_cid);
+    ((void) own_cid_len);
+    return( 0 );
+}
+
+int mbedtls_ssl_get_peer_cid( mbedtls_ssl_context *ssl,
```
Comment (mpg) [inline library/ssl_tls.c]: Please add a big fat warning to the documentation of the function that so far it does nothing, except always set `*enabled` to `DISABLED`.

File: include/mbedtls/config.h [hunk 12]
```diff
@@ -1269,6 +1269,29 @@
...
+ * allows to identify DTLS connections across changes in the
+ * underlying transport.
+ *
+ * Setting this option enables the SSL APIs `mbedtls_ssl_set_cid()`
+ * and `mbedtls_ssl_get_peer_cid()`. See their documentation for more
+ * information.
+ *
+ * \warning The Connection ID extension is still in draft state.
+ *          We make no stability promises for the availability
+ *          or the shape of the API controlled by this option.
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/config.h]: Agreed, will add this.

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Agreed.

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Oh, very good point! To be honest, I haven't thought about that. I think it should apply to all subsequent handshakes unles changed. Are you ok with that?

Comment (mpg) [reply include/mbedtls/ssl.h]: Yeah, I think that's the behaviour I would expect as well.

File: include/mbedtls/check_config.h (hunk 3 above)
Comment (hanno-arm) [reply include/mbedtls/check_config.h]: Yes, though that assumes I don't forget about it, which is actually what happened most times so far. The splitting is never intentional.

File: programs/ssl/ssl_server2.c (hunk 4 above)
Comment (hanno-arm) [reply programs/ssl/ssl_server2.c]: Ok.

File: programs/ssl/ssl_server2.c (hunk 5 above)
Comment (hanno-arm) [reply programs/ssl/ssl_server2.c]: Ok.

File: programs/ssl/ssl_client2.c (hunk 6 above)
Comment (hanno-arm) [reply programs/ssl/ssl_client2.c]: Ok.

File: tests/ssl-opt.sh (hunk 7 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Ops... fixing.

File: tests/ssl-opt.sh (hunk 8 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: The CID influences record protection in a way depending on the method of encryption, so I think it's vital to test AEAD vs. CBC explicitly.

File: tests/ssl-opt.sh (hunk 9 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Yes.

File: library/ssl_tls.c (hunk 10 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Yes. The PR was originally intended to be merged only when the implementation is complete, but as we discussed, this is no longer the case, and I agree with your remark in that case.

File: library/ssl_tls.c (hunk 11 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Yes - as above.

File: include/mbedtls/check_config.h (hunk 3 above)
Comment (mpg) [reply include/mbedtls/check_config.h]: Yeah, I know, same for me. Still think it would be good for all of us to try to create that mental association "extend config.h" -> "re-generate files and consider checks".

File: tests/ssl-opt.sh (hunk 8 above)
Comment (mpg) [reply tests/ssl-opt.sh]: Ok, that's a good reason, I wasn't aware of that.

Comment (hanno-arm) [reply tests/ssl-opt.sh]: Another reason is that we'll likely implement AEAD before CBC, and then we want to test that CID works for AEAD ciphersuites, while the CBC tests will continue to be stubs.

File: tests/ssl-opt.sh [hunk 13]
```diff
@@ -1112,6 +1112,103 @@ run_test    "Truncated HMAC, DTLS: client enabled, server enabled" \
             -S "dumping 'expected mac' (20 bytes)" \
             -s "dumping 'expected mac' (10 bytes)"
 
+# Tests for DTLS Connection ID extension
+
+# TODO
+# So far, the CID API isn't implemented, so we can't
+# grep for output witnessing its use. This needs to be
+# changed once the CID extension is implemented.
```
Comment (mpg) [inline tests/ssl-opt.sh]: [I tried leaving that comment before but now I can't see it anywhere, so leaving it again] I'd much rather have the tests be complete and skipped than incomplete, for two reasons:

1. Incomplete tests diminishes the value of writing them and reviewing them now.
2. As long as the feature isn't actually implemented, I don't think tests should print "PASS" as that would be misleading - "SKIP" on the other hand is much more truthful.

Comment (hanno-arm) [reply tests/ssl-opt.sh]: I am not sure about this, because we can gradually make the tests more expressive. Specifically: (a) Once the new CID handshake fields are in place and configurable through the top-level API, their setting can be witnessed by a debug message we can grep for. (b) Once the CID extension writing/parsing is in place, we can witness that in a debug message we can grep for. (c) Once the CID record protection is in place, we can witness that in a debug message we can grep for.
So, in summary, I expect more and more debug output to be grepped for here, and already the intermediate stages have value.

Comment (mpg) [reply tests/ssl-opt.sh]: You're making excellent points here, I hadn't considered that strategy.

I'm still worried that the output of the script might be misleading. Can we change the description of individual tests then, adding something like "STUB" or "WORK IN PROGRESS - extension sent and received" (then update that at each step until we can remove the WIP part altogether)?

File: include/mbedtls/config.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/config.h]: Done.

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Done.

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Done.

File: programs/ssl/ssl_server2.c (hunk 4 above)
Comment (hanno-arm) [reply programs/ssl/ssl_server2.c]: Done.

File: programs/ssl/ssl_server2.c (hunk 5 above)
Comment (hanno-arm) [reply programs/ssl/ssl_server2.c]: Done.

File: programs/ssl/ssl_client2.c (hunk 6 above)
Comment (hanno-arm) [reply programs/ssl/ssl_client2.c]: Done.

File: tests/ssl-opt.sh (hunk 7 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Fixed.

File: tests/ssl-opt.sh (hunk 9 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Fixed.

File: library/ssl_tls.c (hunk 10 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Done.

File: library/ssl_tls.c (hunk 11 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Done.

File: tests/ssl-opt.sh (hunk 13 above)
Comment (hanno-arm) [reply tests/ssl-opt.sh]: Good idea, I'll do that.

Comment (hanno-arm) [reply tests/ssl-opt.sh]: Done.

**Comment by hanno-arm** [comment]
@mpg Thank you for your feedback, I believe to have addressed your comments. Could you please re-review?

File: include/mbedtls/check_config.h (hunk 3 above)
Comment (mpg) [reply include/mbedtls/check_config.h]: (Also, if we think about it before we pushed (or at least before review started), we can squash the commits together for a cleaner history.)

File: programs/ssl/ssl_server2.c [hunk 14]
```diff
@@ -2876,7 +2876,16 @@ int main( int argc, char *argv[] )
         }
         else
         {
+            size_t idx=0;
             mbedtls_printf( "Use of Connection ID has been negotiated.\n" );
+            mbedtls_printf( "Peer CID (length %u Bytes): ",
+                            (unsigned) peer_cid_len );
+            while( idx < peer_cid_len )
```
Comment (mpg) [inline programs/ssl/ssl_server2.c]: I'd have expected a `for` loop here, but OK.

File: include/mbedtls/check_config.h (hunk 3 above)
Comment (jarlamsa) [reply include/mbedtls/check_config.h]: Would be nice to collect these way of working comments somewhere so they don't get lost.

File: library/ssl_tls.c [hunk 15]
```diff
@@ -109,6 +109,39 @@ static void ssl_update_in_pointers( mbedtls_ssl_context *ssl,
 
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
 
+/* Top-level Connection ID API */
```
Comment (mpg) [inline library/ssl_tls.c]: Found by the CI (test-ref-configs): the definition should be guarded by the same `#ifdef`s as the declaration.

File: programs/ssl/ssl_client2.c [hunk 16]
```diff
@@ -536,6 +551,45 @@ int idle( mbedtls_net_context *fd,
...
+    if( len % 2 != 0 )
+        return( -1 );
+
+    for( j = 0; j < len; j += 2 )
+    {
+        c = hex[j];
+        if( c >= '0' && c <= '9' )
+            c -= '0';
+        else if( c >= 'a' && c <= 'f' )
+            c -= 'a' - 10;
```
Comment (jarlamsa) [inline programs/ssl/ssl_client2.c]: This is a bit confusing, I am not sure what is the desired output now, everything between 0xa and 0xf would be reduced to 0-5.

File: library/ssl_tls.c [hunk 17]
```diff
@@ -109,6 +109,39 @@ static void ssl_update_in_pointers( mbedtls_ssl_context *ssl,
...
+    ((void) own_cid_len);
+    return( 0 );
+}
+
+/* WARNING: This implementation is a stub and doesn't do anything!
+ *          It is included solely to allow review and coding against
+ *          the new Connection CID API. */
+int mbedtls_ssl_get_peer_cid( mbedtls_ssl_context *ssl,
+                     int *enabled,
+                     unsigned char peer_cid[ MBEDTLS_SSL_CID_OUT_LEN_MAX ],
```
Comment (jarlamsa) [inline library/ssl_tls.c]: Should this be of size `MBEDTLS_SSL_CID_IN_LEN_MAX` ?

File: library/ssl_tls.c (hunk 15 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Done.

File: programs/ssl/ssl_client2.c (hunk 16 above)
Comment (hanno-arm) [reply programs/ssl/ssl_client2.c]: @jarlamsa I'm not sure I follow - the logic is the same as `c = ( c - 'a' ) + 10`, which seems correct to me.

File: library/ssl_tls.c (hunk 17 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: @jarlamsa No, the peer CID is the one we use for messages _to_ the peer, while our own CID is what we want the peer to use when it sends messages to us. Does this make sense?

**Comment by hanno-arm** [comment]
@jarlamsa Thank you for your review! I believe I have addressed your comments. Could you take another look?

**Comment by hanno-arm** [comment]
Travis failed because of bad UDP Proxy seed (known issue). Restarting.

File: programs/ssl/ssl_client2.c (hunk 16 above)
Comment (jarlamsa) [reply programs/ssl/ssl_client2.c]: Sorry, missed the double minus. 
But the refactoring comment is still valid although minor.

Comment (hanno-arm) [reply programs/ssl/ssl_client2.c]: @jarlamsa I agree, and we have considered reducing the overlap of `ssl_client2` and `ssl_server2` in other places, too. However, doing that is outside of the scope of this PR. I think @gilles-peskine-arm has done some work in that direction already (?).

Comment (gilles-peskine-arm) [reply programs/ssl/ssl_client2.c]: @hanno-arm You must be thinking of https://github.com/ARMmbed/mbedtls/pull/1272 where I proposed to introduce a `ssl_test_lib.c` used by both `ssl_client2` and `ssl_server2`. That work is currently stalled.

**Comment by mpg** [comment]
@sbutcher-arm This PR now has two approvals and a passing CI, please consider it for merging.

**Comment by mpg** [comment]
Ping @sbutcher-arm for gatekeeping.

File: include/mbedtls/ssl.h [hunk 18]
```diff
@@ -1396,6 +1410,135 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+ *                    by the application layer and are exchanged in new
+ *                    `ClientHello` / `ServerHello` extensions during the
+ *                    handshake, where each side indicates the CID it wants the
+ *                    peer to use when writing encrypted messages. The CIDs are
+ *                    put to use once records get encrypted: the stack discards
+ *                    any incoming records that don't include the configured CID
+ *                    in their header, and adds the peer's requested CID to the
+ *                    headers of outgoing messages.
+ *
+ *                    This API allows to enable/disable the use of the CID
```
Comment (sbutcher-arm) [inline include/mbedtls/ssl.h]: Again, this doesn't work grammatically. How about:
```
This API allows the enabling and disabling of the use of the CID...
```

Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Ok, let's fix it in one of the remaining CID PRs.

File: include/mbedtls/ssl.h [hunk 19]
```diff
@@ -1396,6 +1410,135 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+/**
+ * \brief             (STUB) Configure the use of the Connection ID (CID)
+ *                    extension in the next handshake.
+ *
+ *                    Reference:
+ *                    https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-04
+ *
+ *                    The DTLS CID extension allows to reliably associate
+ *                    DTLS records to DTLS connections across changes in the
+ *                    underlying transport (changed IP+Port metadata) by adding
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Ok, let's fix it in one of the remaining CID PRs.

File: include/mbedtls/config.h [hunk 20]
```diff
@@ -1269,6 +1269,30 @@
  */
 #define MBEDTLS_SSL_ALL_ALERT_MESSAGES
 
+/**
+ * \def MBEDTLS_SSL_CID
+ *
+ * Enable support for the DTLS Connection ID extension
+ * (version draft-ietf-tls-dtls-connection-id-04)
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/config.h]: Ok!

File: include/mbedtls/ssl.h [hunk 21]
```diff
@@ -1396,6 +1410,135 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
+
+#if defined(MBEDTLS_SSL_CID)
+
+
+/**
+ * \brief             (STUB) Configure the use of the Connection ID (CID)
+ *                    extension in the next handshake.
+ *
+ *                    Reference:
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Ok!

**Comment by hanno-arm** [comment]
@sbutcher-arm Thanks for your review! As you suggest, I prefer to implement the minor changes you suggested in one of the follow-up PRs.

File: include/mbedtls/ssl.h [hunk 22]
```diff
@@ -1396,6 +1410,135 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+ *                    https://tools.ietf.org/html/draft-ietf-tls-dtls-connection-id-04
+ *
+ *                    The DTLS CID extension allows to reliably associate
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Ok!
